### PR TITLE
[rocdl] Create basic code structure for ROCDL CodeGen

### PR DESCRIPTION
**Pull request by antiagainst**
This commit starts the rework of LLVMGPU compiler backend for future production. Specifically it starts a set of new source code files for ROCDL CodeGen configuration. The goal is to start with a minimal working structure and gradually port over relevant pass pipelines with proper target handling and heuristics. So in this commit we only pick the warp reduction pipeline which sort of matches the criteria here. Also we don't shuffle the passes around yet--the new ROCDL library just depends on the existing LLVMGPU library and its passes directly to get started. This is meant to avoid churns and we can do the mechanical change once everything is sorted.

Progress towards https://github.com/openxla/iree/issues/16341

### CODE REVIEW

**Review by raikonenfnu — COMMENTED** [review]
Is this is not integrated into e2e iree-compile yet?

**Comment by antiagainst** [comment]
No this is detached from the normal compilation flow now and will remain so until we have cleaned up and ported over all rocdl related pass pipelines. And then we switch the rocm target to use this new strategy selection and codegen pipeline. See detailed steps in https://github.com/openxla/iree/issues/16341. This means we can better isolate the two worlds and balance short-/long-term development goals and have clear cut boundaries regarding certain cleanups (like the KernelConfig etc.)

File: compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  // Decompose affine ops.
+  pm.addPass(createDecomposeAffineOpsPass());
+  // Get rid of the redundant computations.
+  pm.addPass(createCSEPass());
+  // Hoist the resulting decompositions.
+  pm.addPass(createLoopInvariantCodeMotionPass());
+  pm.addPass(createLowerAffinePass());
+}
+
+static void addLowerToROCDLPasses(OpPassManager &pm, bool useROCM) {
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp]: NIT: Probably don't need these useROCM anymore

File: compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+//===----------------------------------------------------------------------===//
+
+void buildROCDLCodegenConfigurationPassPipeline(OpPassManager &pm) {
+  addCommonTargetExecutablePreprocessingPasses(pm);
+  auto &nestedModulePM = pm.nest<ModuleOp>();
+  nestedModulePM.addNestedPass<func::FuncOp>(createGPUGeneralizeNamedOpsPass());
+  pm.addPass(createROCDLSelectLoweringStrategyPass());
+}
+
+void buildROCDLCodegenPassPipeline(OpPassManager &pm, bool useROCM) {
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp]: NIT: same as above

**Review by raikonenfnu — APPROVED** [review]
Thanks for setting this up, overall LGTM! some minor NITs.

**Comment by raikonenfnu** [comment]
> No this is detached from the normal compilation flow now and will remain so until we have cleaned up and ported over all rocdl related pass pipelines. And then we switch the rocm target to use this new strategy selection and codegen pipeline. See detailed steps in #16341. This means we can better isolate the two worlds and balance short-/long-term development goals and have clear cut boundaries regarding certain cleanups (like the KernelConfig etc.)

Understood :)

File: compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+//===----------------------------------------------------------------------===//
+
+// Add passes to make the address computation more explicit and optimize them.
+//
+// The idea here is to be less dependent on what the LLVM backend is able to do,
+// by heavy lifting most of the work while we still have the information about
+// loops.
+//
+// Note that this needs to run before SCF -> CF.
+static void addLowerAndOptimizeAddressComputationPasses(OpPassManager &pm) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp]: THere is a lot of duplication here. This is exactly what the CUDA side does? I am OK with some replication, but these seem like generic utilities.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
It seems like we are branching the whole LLVMGPU compiler. I am not sure if that kind of fracturing is worth it. Could we get an idea of why this fracturing is needed? Somethings I understand is 
1) We need to set different tile sizes, etc. But just like on the SPIR-V side for different vendors you have different tile sizes picked, same should work here
2) Do we need separate lowering pipelines for ROCDL as opposed to what we've been doing in LLVMGPU.

Or is the plan that we start from clean slate for ROCDL, and deprecate LLVMGPU?

**Comment by antiagainst** [comment]
The goal of having ROCDL subdirectory is indeed to start from a clean state so we can port over pieces gradually. I find it super hard to untangle while maintaining functionality / perf not regressing in the current working e2e compiler path. Better to separate and avoid adding more conditions and branches in the current flow to differentiate alpha vs. beta state development further.

Regarding duplication, I anticipate only duplications over policy parts, like the KernelConfig, LowerExecutableTarget pass switch statement, etc. The majority of the code like passes and pipelines should be common (with target description accepted to control the CodeGen). That is reflected with the current code change, where you see only a few passes are copied over and cleaned up. It still relies on LLVMGPU for all the concrete passes. So in the end state I won't expect to see a lot of duplication in general. Even for the policy parts, we can also unify if we, say, can find common heuristics for both. I'd like to avoid shuffling code too much or unify too early until we cleaned up ROCDL and NVVM. Then we can move passes properly to some common directory.

I've listed some more detailed backgrounds / plans / steps in https://github.com/openxla/iree/issues/16341. Let's also touch base on this in the next CodeGen sync. :)

File: compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp (hunk 3 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/ROCDL/Passes.cpp]: Changed to reuse existing code.

**Comment by antiagainst** [comment]
Okay as chatted offline, for now we will still keep all the source files in the same lLVMGPU directory to get started. I've revised the code to do that.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLKernelConfig.cpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+struct TargetInfo {
+  bool hasWarpShuffle = false;
+  // These are listed in the order of preference, not necessarily monotonically.
+  SmallVector<int64_t, 2> supportedSubgroupSizes = {32};
+};
+
+static StringRef getTargetArch(mlir::FunctionOpInterface entryPoint) {
+  if (auto variantOp =
+          entryPoint->getParentOfType<IREE::HAL::ExecutableVariantOp>()) {
+    IREE::HAL::ExecutableTargetAttr targetAttr = variantOp.getTarget();
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLKernelConfig.cpp]: nit: use `IREE::HAL::ExecutableTargetAttr::lookup(entryPoint)`, no need to search for the variant parent first anymore.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLKernelConfig.cpp [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  if (auto linalgOp = dyn_cast<linalg::LinalgOp>(computeOp)) {
+    if (succeeded(setWarpReductionConfig(entryPointFn, linalgOp, targetInfo))) {
+      return success();
+    }
+  }
+
+  return failure();
+}
+
+// Propogates the configuration to the other ops.
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLKernelConfig.cpp]: typo nit: Propagates

**Review by qedawkins — APPROVED** [review]
Nice first step, new pipeline looks so clean :D

**Review by MaheshRavishankar — APPROVED** [review]
Thanks @antiagainst . This seems much better to me.

**Pull request merged**